dep0/lib/03_typecheck/src/check.cpp

```diff
@@ -315,13 +316,37 @@ check_stmt(
                 {
                     std::ostringstream err;
                     pretty_print(err << "expecting expression of type `", state.goal) << '`';
-                    return error_t::from_error(dep0::error_t(err.str(), s.properties), env, state.context, state.goal);
+                    return error_t::from_error(dep0::error_t(err.str(), loc), env, state.context, state.goal);
                 }
             }
             else if (auto expr = check_expr(env, state.context, *x.expr, state.goal, is_mutable, usage, usage_multiplier))
                 return make_legal_stmt(stmt_t::return_t{std::move(*expr)});
             else
                 return std::move(expr.error());
+        },
+        [&] (parser::stmt_t::impossible_t const& x) -> expected<stmt_t>
+        {
+            auto const check_impossible_stmt = [&] (ctx_t const& ctx) -> expected<stmt_t>
+            {
+                auto const false_type = sort_t{derivation_rules::make_true_t(derivation_rules::make_false())};
+                for (auto const& v: ctx.vars())
+                    if (is_beta_delta_equivalent(env, ctx, ctx[v]->value.type, false_type))
+                        return make_legal_stmt(stmt_t::impossible_t{});
```

Comment:
If available, the reason should be added to the legal statement being generated here.